File: appLab4B/src/main/java/com/github/hummel/cga/lab4b/Matrix4.java
package com.github.hummel.cga.lab4b;

public class Matrix4 {
    private final double[][] data;

    public Matrix4() {
        data = new double[][]{
                {1.0, 0.0, 0.0, 0.0},
                {0.0, 1.0, 0.0, 0.0},
                {0.0, 0.0, 1.0, 0.0},
                {0.0, 0.0, 0.0, 1.0}
        };
    }

    public void set(int i, int j, double val) {
        data[i][j] = val;
    }

    public Matrix4 mul(Matrix4 arg) {
        var res = new Matrix4();
        for (var i = 0; i < 4; i++) {
            for (var j = 0; j < 4; j++) {
                res.data[i][j] = 0.0;
                for (var k = 0; k < 4; k++) {
                    res.data[i][j] += data[i][k] * arg.data[k][j];
                }
            }
        }
        return res;
    }

    public Vector4 mul(Vector4 arg) {
        var res = new Vector4();
        res.set(3, 0.0);
        for (var i = 0; i < 4; i++) {
            for (var j = 0; j < 4; j++) {
                res.set(i, res.get(i) + data[i][j] * arg.get(j));
            }
        }
        return res;
    }
}

File: appLab4B/src/main/kotlin/com/github/hummel/cga/lab4b/Matrix4.kt
package com.github.hummel.cga.lab4b

class Matrix4 {
	private val data = arrayOf(
		doubleArrayOf(1.0, 0.0, 0.0, 0.0),
		doubleArrayOf(0.0, 1.0, 0.0, 0.0),
		doubleArrayOf(0.0, 0.0, 1.0, 0.0),
		doubleArrayOf(0.0, 0.0, 0.0, 1.0)
	)

	fun set(i: Int, j: Int, `val`: Double) {
		data[i][j] = `val`
	}

	fun mul(arg: Matrix4): Matrix4 {
		val res = Matrix4()
		for (i in 0..3) {
			for (j in 0..3) {
				res.data[i][j] = 0.0
				for (k in 0..3) {
					res.data[i][j] += data[i][k] * arg.data[k][j]
				}
			}
		}
		return res
	}

	fun mul(arg: Vector4): Vector4 {
		val res = Vector4()
		res[3] = 0.0
		for (i in 0..3) {
			for (j in 0..3) {
				res[i] = res[i] + data[i][j] * arg[j]
			}
		}
		return res
	}
}
